### ISSUE TITLE
VIM not compiling with Python 3.10, but did previously with same env and Python 3.9

### ISSUE THREAD

**Issue by broander** [vim/vim#12184]
Hi,

I've summarized the entire issue in a VIM Stack Overflow question here that hasn't yielded any responses: https://vi.stackexchange.com/questions/41601/compiling-vim-from-source-with-python3-enabled-not-working-after-upstream-image

Here's the config command I'm running to compile VIM (more details in the article linked above):

./configure --enablepython3interp=yes --disable-gui --without-x --enable-gui=no

I've found the line in the config.log that is failing, it's the following:

configure:6947: checking if compile and link flags for Python 3 are sane
configure:6964: gcc -o conftest -g -O2 -I/opt/conda/include/python3.10 -pthread   -L/usr/local/lib conftest.c  -lselinux  -L/opt/conda/lib/python3.10/config-3.10-x86_64-linux-gnu -lpython3.10 -lcrypt -lpthread -ldl -lutil -lrt -lm -lm >&5
/usr/bin/ld: cannot find -lpython3.10
collect2: error: ld returned 1 exit status

Would anyone here be familiar with how to modify my configure command to make this work, or is this a bug?  I'm concerned it might be a bug, because the same reproducible codespace environment would compile VIM under Python 3.9, but now doesn't under Python 3.10 after the upstream image updated to Python 3.10.

Also, I've checked the config.log in another codespace I have from several weeks ago that's still running Python 3.9, and here's the output without the error:

configure:6947: checking if compile and link flags for Python 3 are sane
configure:6964: gcc -o conftest -g -O2 -I/opt/conda/include/python3.9 -pthread   -L/usr/local/lib conftest.c  -lselinux  -L/opt/conda/lib/python3.9/config-3.9-x86_64-linux-gnu -lpython3.9 -lcrypt -lpthread -ldl -lutil -lm -lm >&5
configure:6964: $? = 0
configure:6965: result: yes

Full output of the config.log with the error is also included in the link above.

Thanks for your help!

**Comment by chrisbra** [comment]
why did you not fill out the issue template? We don't know what version of Vim you used. Also, you may want to test, if it works with regular python and not conda. Can you please check that?

**Comment by ychin** [comment]
Did you try to locate the Python library to see if it exists? The compiler error message is saying that it is having issues loading the lib, which it is looking for under `/opt/conda/lib/python3.10/config-3.10-x86_64-linux-gnu`. The compiler command is basically telling you most information you should need, with `-L` specifying the folder it's looking under, and `-lpython3.10` meaning it's looking for a "libpython3.10.a".

**Comment by broander** [comment]
Thanks @ychin for the hint--the libpython file was indeed not in the location that the compiler was trying to find it in.  So nearest I can figure is this wasn't an issue when the system python and conda python were both 3.9, because the versions matched in the check process and passed.  When the upstream imaged rolled conda python to 3.10, they no longer matched and failed.

Using the find-libpython module to track it down (because "find" wasn't cutting it), I found that the conda env lib is '/opt/conda/lib/libpython3.10.so.1.0'

So, I tried adding this argument to the config command: '--with-python3-config-dir=/opt/conda/lib'.

It worked!  The config output now said that the compile and link flags are sane.

So, I ran make and then installed.  And..... it didn't work.  VIM crashed with this as the error:
vim: error while loading shared libraries: libpython3.10.so.1.0: cannot open shared object file: No such file or directory

I believe that the '--with-python3-config-dir=' argument is deprecated, so it might not have been the right path to do down.  Now that we know where the libpython file is located, is there a better argument I can provide configure make?

**Comment by ychin** [comment]
The reason why `--with-python3-config-dir=` is deprecated is because the configure script tries to derive that automatically using Python itself. It uses whatever `python3` is in your PATH, or you can use `--with-python3-command=/my/path/python3`. It basically uses that Python and calls the following, hoping to find a folder that contains the libs:

```
import sysconfig; print(sysconfig.get_config_var('LIBPL'))
```

However, it seems that conda does not respect that convention and put it somewhere else, so you do have to manually set the config dir.

> So nearest I can figure is this wasn't an issue when the system python and conda python were both 3.9, because the versions matched in the check process and passed

> It worked! The config output now said that the compile and link flags are sane.

> So, I ran make and then installed. And..... it didn't work. VIM crashed with this as the error:
vim: error while loading shared libraries: libpython3.10.so.1.0: cannot open shared object file: No such file or directory

I think what is happening here is that the conda Python lib is doing a bunch of stuff that pulls in other dynamic libraries that don't exist on your system, and yes, I believe it used to just use the system Python 3.9 libs. If you do `ldd <path_to_vim>` it will tell you all the dependencies the binary has and you will probably be able to spot the missing libs that it's trying to load.

What you probably want to do is to actually link against Python using dynamic linkage. So modify your configure script to do this instead (`vi_cv_dll_name_python3` is an env var that's not exposed through configure script command-line params):

```
vi_cv_dll_name_python3=/opt/conda/lib/libpython3.10.so.1.0 ./configure --enable-python3interp=dynamic <the other flags you were using>
```

You can also look at the compiler flags the configure script generates by looking at src/auto/config.mk, under variables with "PYTHON3_" prefix.

---

**Before we dive deeper though**, is there a reason why you need the version of Python 3 that Vim embeds to be the exact same version as the conda one and not whatever the system default is? Is there some plugin that require it? Maybe I'm not understanding what you mean by your codespace requiring the Python version to match. The fact that Vim used to work just fine for you even though it's silently using system Python tells me you may not actually need to jump through all these hoops. You can just build Vim with `--with-python3command=/usr/bin/python3` (assuming that's the path of system Python 3.9) and it should just work.

**Comment by broander** [comment]
@ychin Thanks for your continued support!

I just tried compiling with your last suggestion--I agree there's nothing about my requirements that stops us from compiling with the system python (especially because I can still run with whatever interpreter I want to using my .vimrc and plugin setup).  So, I tried this:

./configure --enable-python3interp --disable-gui --without-x --with-python3command=/usr/bin/python3

But, it didn't work..............

Fortunately, easy fix, just needed to fix the typo of the new command and add a dash=)

./configure --enable-python3interp --disable-gui --without-x --with-python3-command=/usr/bin/python3

And that worked!  VIM compiled with Python3 enabled, and starts up like a champ.  That works fine, no need to go deeper and figure out how to link everything to the separate env that conda installed.  Thanks for your help!
